backend/gs/file.go

```diff
@@ -25,84 +25,304 @@ const (
 
 // File implements vfs.File interface for GS fs.
 type File struct {
-	fileSystem  *FileSystem
-	bucket      string
-	key         string
-	tempFile    *os.File
-	writeBuffer *bytes.Buffer
+	fileSystem *FileSystem
+	bucket     string
+	key        string
+
+	// seek-related fields
+	cursorPos  int64
+	seekCalled bool
+
+	// read-related fields
+	reader     io.ReadCloser
+	readCalled bool
+
+	// write-related fields
+	tempFileWriter *os.File
+	gcsWriter      io.WriteCloser
+	cancelFunc     context.CancelFunc
+	writeCalled    bool
 }
 
 // Close cleans up underlying mechanisms for reading from and writing to the file. Closes and removes the
-// local temp file, and triggers a write to GCS of anything in the f.writeBuffer if it has been created.
+// local temp file, and triggers a Write to GCS of anything in the f.writeBuffer if it has been created.
 func (f *File) Close() error {
-	if f.tempFile != nil {
-		defer func() { _ = f.tempFile.Close() }()
+	defer func() {
+		f.reader = nil
+		f.cancelFunc = nil
+		f.gcsWriter = nil
+
+		// reset state
+		f.cursorPos = 0
+		f.seekCalled = false
+		f.readCalled = false
+		f.writeCalled = false
+	}()
 
-		err := os.Remove(f.tempFile.Name())
-		if err != nil && !os.IsNotExist(err) {
-			return err
+	// cleanup reader (unless reader is also the writer tempfile)
+	if f.reader != nil && !f.writeCalled {
+		// close reader
+		if err := f.reader.Close(); err != nil {
+			return utils.WrapCloseError(err)
 		}
+	}
 
-		f.tempFile = nil
+	// finalize writer
+	if f.gcsWriter != nil {
+		// close s3Writer
```

Comment:
```suggestion
		// close gcsWriter
```

Reply:
Fixed. Obviously this was largely a cut/paste from the s3 implemenation.